Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/MainActivity.java

```diff
@@ -113,8 +113,12 @@ private void analyze() {
     totalFrames =
         Integer.valueOf(mediaMetadataRetriever.extractMetadata(METADATA_KEY_VIDEO_FRAME_COUNT));
     // TODO: Add on demand frame read opposed to read all frames.
-    frameList = mediaMetadataRetriever.getFramesAtIndex(0, Math.max(0, totalFrames - SAFE_FRAMES));
-
+    for (int i = 0; i < totalFrames; i += FRAME_CHUNK_READ_SIZE) {
+      if (i + FRAME_CHUNK_READ_SIZE - 1 < totalFrames)
+        frameList.addAll(mediaMetadataRetriever.getFramesAtIndex(i, FRAME_CHUNK_READ_SIZE));
+      else
+        frameList.addAll(mediaMetadataRetriever.getFramesAtIndex(i, totalFrames-i));
```

Comment:
frameList.addAll(mediaMetadataRetriever.getFramesAtIndex(i, Math.min(totalFrames - i, FRAME_CHUNK_READ_SIZE))

Reply:
That makes sense. Thanks!
> Looks ok if it works, but a better design would be to only read chunks of frames as we need them for analysis, and then discard them after the analysis to reduce memory usage.

I added a TODO to read frames on demand, I planned to do that when optimizing the linear algorithm for finding `serverSequence` among the OCR texts.